File: crates/headgate/src/insert_hook.rs
//! Non-wrapping observers around each actual enqueue store attempt.
//!
//! Middleware owns control flow and may call its inner chain zero, one, or many times.
//! Insert hooks do not receive `next`: every downstream call that reaches the store emits
//! exactly one `Begin` and one `End` event, in registration order at both phases.

use headgate_core::{Envelope, StoreError};

use crate::{EnqueueContext, EnqueueOperation};

/// The immutable request snapshot immediately before the selected store method is called.
/// Hooks cannot rewrite the request; mutation belongs in enqueue middleware.
#[derive(Clone, Copy, Debug)]
pub struct InsertAttempt<'a> {
    context: &'a EnqueueContext,
    operation: EnqueueOperation,
    batch: &'a [Envelope],
}

impl<'a> InsertAttempt<'a> {
    pub(crate) fn new(
        context: &'a EnqueueContext,
        operation: EnqueueOperation,
        batch: &'a [Envelope],
    ) -> Self {
        Self {
            context,
            operation,
            batch,
        }
    }

    pub fn context(self) -> &'a EnqueueContext {
        self.context
    }

    pub fn operation(self) -> EnqueueOperation {
        self.operation
    }

    pub fn batch(self) -> &'a [Envelope] {
        self.batch
    }
}

/// The store result observed by an insert-end hook.
///
/// `Succeeded` includes a new insert and an idempotent same-id replay because the Store
/// port deliberately returns `()` for both. Duplicate and id conflict remain explicit;
/// every other reachable-store rejection retains the original typed error.
#[derive(Clone, Copy, Debug)]
pub enum InsertOutcome<'a> {
    Succeeded,
    Duplicate {
        existing_id: &'a str,
        replaced: bool,
    },
    IdConflict {
        job_id: &'a str,
    },
    Rejected {
        error: &'a StoreError,
    },
}

impl<'a> InsertOutcome<'a> {
    pub(crate) fn from_store_result(result: &'a Result<(), StoreError>) -> Self {
        match result {
            Ok(()) => Self::Succeeded,
            Err(StoreError::Duplicate {
                existing_id,
                replaced,
            }) => Self::Duplicate {
                existing_id,
                replaced: *replaced,
            },
            Err(StoreError::IdConflict { job_id }) => Self::IdConflict { job_id },
            Err(error) => Self::Rejected { error },
        }
    }

    pub fn is_succeeded(self) -> bool {
        matches!(self, Self::Succeeded)
    }
}

/// One point event in the insert lifecycle. Hooks are called in registration order for
/// both phases; end hooks do not unwind like middleware.
#[derive(Clone, Copy, Debug)]
pub enum InsertHookEvent<'a> {
    Begin {
        attempt: InsertAttempt<'a>,
    },
    End {
        attempt: InsertAttempt<'a>,
        outcome: InsertOutcome<'a>,
    },
}

impl<'a> InsertHookEvent<'a> {
    pub fn attempt(self) -> InsertAttempt<'a> {
        match self {
            Self::Begin { attempt } | Self::End { attempt, .. } => attempt,
        }
    }

    pub fn outcome(self) -> Option<InsertOutcome<'a>> {
        match self {
            Self::Begin { .. } => None,
            Self::End { outcome, .. } => Some(outcome),
        }
    }
}

/// A synchronous, non-wrapping insert lifecycle observer.
///
/// Hooks should hand expensive work to an asynchronous exporter. They cannot veto,
/// mutate, retry, or replace a store result; use [`crate::EnqueueMiddleware`] for those
/// control-flow responsibilities.
pub trait InsertHook: Send + Sync + 'static {
    fn on_insert(&self, event: InsertHookEvent<'_>);
}

/// Function adapter for a lightweight insert hook.
pub struct InsertHookFn<F>(F);

impl<F> InsertHookFn<F> {
    pub fn new(function: F) -> Self {
        Self(function)
    }
}

impl<F> InsertHook for InsertHookFn<F>
where
    F: for<'a> Fn(InsertHookEvent<'a>) + Send + Sync + 'static,
{
    fn on_insert(&self, event: InsertHookEvent<'_>) {
        (self.0)(event);
    }
}

File: go/insert_hook.go
package headgate

import (
	"context"
	"errors"
)

// InsertAttempt is an immutable snapshot immediately before an enqueue store call.
// Batch returns a deep copy so a hook cannot mutate what this or a later hook observes,
// or what the store will persist. Request mutation belongs in enqueue middleware.
type InsertAttempt struct {
	Source    EnqueueSource
	Operation EnqueueOperation
	batch     []Envelope
}

func newInsertAttempt(request EnqueueRequest) InsertAttempt {
	return InsertAttempt{
		Source: request.Source, Operation: request.Operation,
		batch: cloneEnqueueBatch(request.Batch),
	}
}

// Batch returns an independently owned view of the attempted atomic batch.
func (a InsertAttempt) Batch() []Envelope { return cloneEnqueueBatch(a.batch) }

type InsertOutcomeKind string

const (
	// InsertOutcomeSucceeded includes a new insert and an idempotent same-ID replay;
	// Store intentionally returns nil for both.
	InsertOutcomeSucceeded  InsertOutcomeKind = "succeeded"
	InsertOutcomeDuplicate  InsertOutcomeKind = "duplicate"
	InsertOutcomeIDConflict InsertOutcomeKind = "id_conflict"
	InsertOutcomeRejected   InsertOutcomeKind = "rejected"
)

// InsertOutcome preserves the actual store result. Duplicate and ID conflict expose
// their useful identifiers; Err is the original typed error for every non-success.
type InsertOutcome struct {
	Kind       InsertOutcomeKind
	ExistingID string
	Replaced   bool
	JobID      string
	Err        error
}

func classifyInsertOutcome(err error) InsertOutcome {
	if err == nil {
		return InsertOutcome{Kind: InsertOutcomeSucceeded}
	}
	if duplicate, ok := errors.AsType[*DuplicateError](err); ok {
		return InsertOutcome{
			Kind: InsertOutcomeDuplicate, ExistingID: duplicate.ExistingID, Replaced: duplicate.Replaced, Err: err,
		}
	}
	if conflict, ok := errors.AsType[*IDConflictError](err); ok {
		return InsertOutcome{
			Kind: InsertOutcomeIDConflict, JobID: conflict.JobID, Err: err,
		}
	}
	return InsertOutcome{Kind: InsertOutcomeRejected, Err: err}
}

type InsertHookPhase string

const (
	InsertHookBegin InsertHookPhase = "begin"
	InsertHookEnd   InsertHookPhase = "end"
)

// InsertHookEvent is one non-wrapping lifecycle point. Outcome returns false for Begin
// and the exact classified store result for End.
type InsertHookEvent struct {
	phase   InsertHookPhase
	attempt InsertAttempt
	outcome *InsertOutcome
}

func (e InsertHookEvent) Phase() InsertHookPhase { return e.phase }

func (e InsertHookEvent) Attempt() InsertAttempt { return e.attempt }

func (e InsertHookEvent) Outcome() (InsertOutcome, bool) {
	if e.outcome == nil {
		return InsertOutcome{}, false
	}
	return *e.outcome, true
}

// InsertHook observes an actual store attempt without receiving a next function. Hooks
// run in registration order at both phases and cannot veto, mutate, retry, or replace the
// result. Expensive work should be handed to an asynchronous exporter.
type InsertHook interface {
	OnInsert(context.Context, InsertHookEvent)
}

type InsertHookFunc func(context.Context, InsertHookEvent)

func (f InsertHookFunc) OnInsert(ctx context.Context, event InsertHookEvent) {
	f(ctx, event)
}
